services/yellow/src/api/Handlers/ChargeForOrderCancellationFeeHandler.cs

```diff
@@ -0,0 +1,75 @@
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using artiso.AdsdHotel.Yellow.Api.Services;
+using artiso.AdsdHotel.Yellow.Api.Validation;
+using artiso.AdsdHotel.Yellow.Contracts.Commands;
+using artiso.AdsdHotel.Yellow.Contracts.Models;
+using artiso.AdsdHotel.Yellow.Events;
+
+namespace artiso.AdsdHotel.Yellow.Api.Handlers
+{
+    public class ChargeForOrderCancellationFeeHandler : AbstractHandler<ChargeForOrderCancellationFeeRequest, OrderCancellationFeeCharged>
+    {
+        private readonly IOrderService _orderService;
+        private readonly ICreditCardPaymentService _paymentService;
+
+        public ChargeForOrderCancellationFeeHandler(IOrderService orderService,
+            ICreditCardPaymentService paymentService)
+        {
+            _orderService = orderService;
+            _paymentService = paymentService;
+        }
+
+        protected override async Task<OrderCancellationFeeCharged> Handle(ChargeForOrderCancellationFeeRequest message)
+        {
+            var order = await Ensure(message, m => _orderService.FindOneById(m.OrderId));
+
+            var amountToPay = order.Price.CancellationFee;
+
+            if (message.AlternativePaymentMethod is not null)
```

Comment:
I would suggest to separate the retrival of the pm from the charging, like:
```csharp
// 1)
var paymentMethodToUse = message.AlternativePaymentMethod is not null
    ? message.AlternativePaymentMethod
    : await GetLatestPaymentMethod(orderId); // or is it already in: order.OrderPaymentMethods ?

EnsureValid(paymentMethodToUse);

// 2)
await ChargeOrder(order, amount, paymentMethodToUse);
```

Reply:
it seems a good solution, but can't handle like so, because the ChargeOrder  handle the payment for method differently when dealing already used form of payments..